restli-server/src/main/java/com/linkedin/restli/internal/server/response/ResponseUtils.java

```diff
@@ -84,6 +92,95 @@ static Object translateCanonicalKeyToAlternativeKeyIfNeeded(Object canonicalKey,
     }
   }
 
+  /**
+   * Used to fill in the default field for a data record if the record is having the field but not "complete"
+   * Basically, a simplified psuedo code is:
+   *  FillInDefault(schema, dataMap):
+   *   for field in schema.fields:
+   *     fieldSchema = field.getSchema()
+   *     if fieldSchema instanceOf RecordDataTemplate:
+   *       if data.contains(field):
+   *         FillInDefault(fieldSchema, data.get(field))
+   *       else:
+   *         if field.hasDefaultValue():
+   *           DataMap newData = field.getDefault()
+   *           FillInDefault(fieldSchema, newData)
+   *       data.put(field, newData)
+   *     else if fieldSchema instanceOf TypeRefDataTemplate:
+   *       FillIndFefault(schema.deReference(), data)
+   *     else:
+   *       if !data.contains(field) && field.hasDefaultValue():
+   *         data.put(field, field.getDefaultValue())
+   * @param dataSchema -> a schema of the corresponding dataMap
+   * @param dataMap -> a dataMap that needs to be filling the default
+   * @return a new different data map that contains original data plus the default values
+   */
+  public static DataMap fillInDefaultValues(DataSchema dataSchema, DataMap dataMap)
```

Comment:
You would also need a method to handle DataLists. What if you have a field which is an array of records?

Reply:
For fields of array, in the code I assume the following case:
Case 1 When the schema provide default
```
testFieldArray: array[MidLevelRecordWithDefault] = [{"intWithDefault": 0, "intWithoutDefault": 0}]
```
Case 2 When schema does not have default

In case 1, I will just use the default provided.
In case 2, there will be no defaults filled in.

I think this is an important case to discuss.

I think whats the correct way might be: for a field that is an array, if there is no default provided restli provides no value with this field (This statement matches the example I pasted in design doc)

If there is default for an array field, I will just use the default.

I think now the question is, is it just default provided on the schema or the element record all other default values?

Let me know if this is wrong

Reply:
You should follow the same behavior as complex fields.
1) if field is not null, fill it with defaults from item's schema recursively.
2) if field is null and schema provides default, use the default and fill it with defaults from the item's schema recursively.
3) if field is null and not default provided, do nothin.